### ISSUE TITLE
(XML) Highlight !DOCTYPE tag components

### ISSUE THREAD

**Issue by w3suli** [highlightjs/highlight.js#2264]
Currently, the entire member is highlighted with the meta class.
```
<!DOCTYPE html>

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"  "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">

<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1 Basic//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11-basic.dtd">
```

The components could also be highlighted.
My question is, what classes would you use for this?

There are two possible classes between meta-classes. One is the **meta-keyword** for the keywords that mark the document and the other is the **meta-string** for the strings.

Look forward to your suggestions.

**Comment by yyyc514** [comment]
Re XHTML I don’t think this really matters since so much of the web is just HTML 5 now anyways. So this is an edge case to begin with. 

Also not sure it’s a problem this is all meta. Ie, don’t see a lot of value in splitting it out. 

@egor-rogov thoughts?

**Comment by w3suli** [comment]
Highlighting strings would be helpful.
This allows strings to be more separated, if any.
`<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1 Basic//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11-basic.dtd">`
```
<!DOCTYPE note [
<!ENTITY nbsp "&#xA0;">
<!ENTITY writer "Writer: Donald Duck.">
<!ENTITY copyright "Walt Disney Company.">
]>
```
The rest is less important.

**Comment by egor-rogov** [comment]
We make use of `meta-keyword` and `meta-string` in other languages, so I have no objections for them in XML also.

**Comment by yyyc514** [comment]
Well let’s do it then!   I guess I feel all languages are not the same but it hard to turn that into a persuasive argument.

**Comment by w3suli** [comment]
@egor-rogov, @yyyc514 Is this the solution you like?
![doctype](https://user-images.githubusercontent.com/20889555/68788251-53febb80-0643-11ea-9c73-2de9025e5c79.png)

**Comment by yyyc514** [comment]
Why the two different classes?  Tag vs meta?

**Comment by w3suli** [comment]
If doctype is part, then meta, if not part, then not.

```
<!DOCTYPE note
[
<!ELEMENT note (to,from,heading,body)>
<!ELEMENT to (#PCDATA)>
<!ELEMENT from (#PCDATA)>
<!ELEMENT heading (#PCDATA)>
<!ELEMENT body (#PCDATA)>
]>
```
```
<!ELEMENT bookstore (book*)>
<!ELEMENT book (title,author,genre?)>
<!ELEMENT title (#PCDATA)>
<!ELEMENT author (first-name, last-name)>
<!ELEMENT genre (#PCDATA)>
<!ELEMENT first-name (#PCDATA)>
<!ELEMENT last-name (#PCDATA)>
<!ATTLIST book price CDATA #REQUIRED>
<!ATTLIST book publicationdate CDATA>
<!ATTLIST book IBSN CDATA>
```
Should items starting with an exclamation mark be highlighted as meta, even if they are not within the document type?

**Comment by yyyc514** [comment]
Oh I see now... I've never used anything more than DOCTYPE in my entire web development career, so I'm not sure I have the answer here. :-)

I guess when I see `<!SOMETHING` my mind immediate just says "meta" cause they are weird and rare and I've never seen them in the wild, lol... is that not how they are usually classified?

Feels like either might be more correct, but I'm not the person to ask.

**Comment by yyyc514** [comment]
@marcoscaceres Any thoughts here on the appropriate designation?

**Comment by marcoscaceres** [comment]
I’m not sure 🤔 these are “processing directives” and they do important things in browsers and xml in general. But I honestly don’t know what to classify them as. I guess meta is fine?

**Comment by w3suli** [comment]
All right. I'm working on making all "<! .." meta exceptions to the "<! -" note notation.

**Comment by w3suli** [comment]
@egor-rogov, @yyyc514, @marcoscaceres Is this the solution you like?
![doctype2](https://user-images.githubusercontent.com/20889555/68888851-03a75c80-071c-11ea-9621-52cf5fd4a125.png)

**Comment by yyyc514** [comment]
I think that's kind of what I expected it would work like.

**Comment by marcoscaceres** [comment]
Looks good to me.

**Comment by yyyc514** [comment]
Think this is closed by #2287, yes?
